Add a tool to parse multiple files, and return structured data.

morrowc opened the pull request:
Initial commit for irrParser/main.go that uses the rpsl-parser library
to parse many irr files at once and reads from the parser the Records
as they are created.

morrowc committed:
Add a tool to parse multiple files, and return structured data.

Initial commit for irrParser/main.go that uses the rpsl-parser library
to parse many irr files at once and reads from the parser the Records
as they are created.

aashaikh commented:
Would prefer that we use feature branches in this repo, rather than merge from a fork if possible -- makes it easy to see the updated code tree in one place.  Up to you if you'd rather make changes in your fork.

aashaikh commented:
Seems we would want to create `rpsl-tools`, and put both the parser and multi-file reader in there?

morrowc commented:
On Thu, 21 Feb 2019 03:01:35 -0500,
Anees Shaikh <notifications@github.com> wrote:
> 
> [1 <text/plain; UTF-8 (7bit)>] Would prefer that we use feature
> branches in this repo, rather than merge from a fork if possible --
> makes it easy to see the updated code tree in one place.  Up to you
> if you'd rather make changes in your fork.
>

seems fine to me, I don't actually know how to do that :)
my git repertoire is .. more limited than my emacs repertoire :)

> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/manrs-tools/contrib/pull/5#issuecomment-465899980
> [2  <text/html; UTF-8 (7bit)>]

morrowc committed:
Rename the irrParser/ directory to rpsl-tools/

This was, in fact what the original plan was, and my rush to
move forward made me forget the plan :(

morrowc commented:
On Thu, 21 Feb 2019 03:02:38 -0500,
Anees Shaikh <notifications@github.com> wrote:
> 
> Seems we would want to create
> `rpsl-tools`, and put both the parser and multi-file reader in
> there?

I agree that some better organization is necessary.
moved.

> 
> -- 
> You are receiving this because you authored the thread.  Reply to
> this email directly or view it on GitHub:
> https://github.com/manrs-tools/contrib/pull/5#issuecomment-465900312
> [2 <text/html; UTF-8 (7bit)>]

morrowc committed:
Export peek() and use that in the rpsl-tools/main.go

Also, fix tests along the way, of course.

aashaikh commented on rpsl-tools/main.go:
```diff
@@ -0,0 +1,186 @@
+// those to the rpsl-parser library, returning each record to a channel for
+// disposition in the final data structure to be loaded into a DB.
+func main() {
+	flag.Var(&rpslFiles, "rpslFiles", "Files to parse, irr/rpsl content(csv).")
```
does this mean the file contents should be csv, or that the list of files are comma separated?  presumably the latter?

morrowc replied:
list of files is csv.
the actual file content is the 'standard' RPSL backup content.

morrowc replied:
Clarified in the flag guidance.

aashaikh commented on rpsl-tools/main.go:
```diff
@@ -0,0 +1,186 @@
+	}
+
+	// Two buffered channels, one for input and for the resulting records.
+	ic := make(chan string, len(rpslFiles))
```
Is this channel useful here?  Since the list of files is supplied in a flag, would it be simpler to just loop through the specified filenames?

morrowc replied:
I think the parsing parts are seperate goroutines, so to get a filename for processing into them you must have those routines listen to a channel. That's this channel (for input) and the second to receive back here from those parsing routines the records as they appear.

aashaikh commented on rpsl-tools/main.go:
```diff
@@ -0,0 +1,186 @@
+
+		reader, err := getReader(fn)
+		if err != nil {
+			glog.Infof("failed to read file(%v): %v\n", fn, err)
```
nit:  s/failed/Failed to be consistent with other err log messages.

morrowc replied:
done, thanks!

aashaikh commented on rpsl-tools/main.go:
```diff
@@ -0,0 +1,186 @@
+}
+
+// parseFile parses a single file at a time into records sent to the
+// records channel (rc). Files arrive as names on the input channel (ic).
```
May want to add in the comment that an error reading any file aborts reading and moves to the next file.

Also the comment might be a bit confusing in saying that this function parses a single file at time; might say "parseFile reads and parses files as their filenames arrive on the input channel.  A parse error will abort processing for the corresponding file and move to the next one."

morrowc replied:
I like your comment text, thanks!

aashaikh commented on rpsl-tools/main.go:
```diff
@@ -0,0 +1,186 @@
+	}
+
+	// Push each file into the input (ic) channel.
+	for _, fn := range rpslFiles {
```
per the comment above, this patterns seems a bit unintuitive to me, i.e., using an input channel rather than just looping through the files and start a thread to parse each one.  Is the concern that there might be too many files (vis-a-vis num threads)?

morrowc replied:
oh yes, I didn't want to try and start 'too many' routines, perhaps that's overly cautious.
it's a relatively normal pattern i'd seen elsewhere (and used elsewhre).

aashaikh replied:
My understanding is that goroutines are very lightweight (~free) such that it would not be an issue to start one for each filename that you get on input.  I.e., the system will manage for you in terms of multiplexing on OS threads.

morrowc replied:
I'm sure they are lightweight, but you're competing for disk space and cpu and ram, I didn't want to totally crush the box doing the processing and 'as fast as possible' isn't super important here.

aashaikh commented on rpsl-tools/main.go:
```diff
@@ -0,0 +1,186 @@
+	// Track the number of files completed.
+	files := 0
+	// Read records from the channel,
+	// Note: I'm positive this is supposed to be simpler with
```
use `TODO(morrowc)` notation in the comment?

morrowc replied:
done

aashaikh commented on rpsl-tools/main.go:
```diff
@@ -0,0 +1,186 @@
+
+	close(rc)
+
+	// Start goroutines which will read filenames to parse and return records
```
can clean up this commented code?

morrowc replied:
done

aashaikh reviewed (approved):
Left a few minor comments; one question on the input channel pattern for clarification

morrowc committed:
Cleanup of comments/questions from PR#5.

morrowc committed:
Handle a missed comment change request.
